Fix issue with leading zero in version numbers

brouer opened the pull request:
If the numeric parts of either of two dependencies' version strings has
a differing number of leading zeroes, but parse as the same numerical
value, the comparison algorithm breaks out too early.

Signed-off-by: Kim Brouer <kim@baudwalk.com>

Example:
```
V08.02.07.00001
V8.04.00.00091
V8.04.00.00093
V08.04.00.00138
V08.04.00.00139
```
The V08 and V8 mix causes problems.

Only a problem for people with the habit of putting leading zeros in their version numbers, of course.


### Context
<!--- Why do you believe many users will benefit from this change? -->
<!--- Link to relevant issues or forum discussions here -->

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [ ] Recognize contributor in release notes

blindpirate commented:
Thank you @brouer , this looks great. Would you mind adding a unit test for it? It can be in https://github.com/gradle/gradle/blob/489524e3693b542abc4280a4e70a0e5467dd831c/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/strategy/DefaultVersionComparatorTest.groovy#L21

brouer commented:
Done.

I added three tests:

- "compares versions that differ only in leading zeros equal"
- "compares different versions that also differ in leading zeros"
- "compares versions where earlier version parts differ only in leading zeros"

It's actually only the last one that fails without my fix, but I added the other two to make the expected _(by me, anyway)_ behaviour clear.

blindpirate commented:
Thanks @brouer ! @gradle/dependency-management I think this makes sense, WDYT?

melix commented:
I'm not sure we should make this change. There are lots of versioning schemes in the wild and in general, we have to be very soft wrt to expectations. In particular, I don't see why `1.01` and `1.1` should be considered equal. I'm actually pretty sure they should not.

brouer commented:
Actually, 1.01 and 1.1 will already be reported as equal. If they shoudln't, that's another bug.

My problem was that, since the 01 vs 1 comparison will not only report that they are equal, because they are numerically equal, the comparator will also break out of the comparison early.

This results in 1.01.1 and 1.1.2 also being reported as equal, because the comparator never gets to the third part (1 vs 2).

The difference my change makes, is that the comparator does not stop at 01 vs 1, but continues as expected. Deciding that 1.1.2 is larger than 1.01.1.

Leaving 01 equaling 1 was both the easiest, as well as the minimal change in behavior to fix the bug.
If 01 and 1 should not be considered equal, we would at least have to decide which is the larger of the two.

blindpirate commented:
@melix can you elaborate on why `1.01` and `1.1` should not be equal? IMO they're equal.

@brouer I also wonder what's the user case, or motivation to drive you make such change.

Anyway, I'll leave this to dependency management team to decide.

brouer commented:
Sorry I haven't been clear.

I'm **not** trying to change existing behaviour.
I actually have no firm opinion on whether "1.01" and "1.1" should be equal or not. But the existing implementation **does** consider them equal. My patch doesn't change that.

The unit tests "compares versions that differ only in leading zeros equal" and "compares different versions that also differ in leading zeros" pass with the existing code as well as with my patch.
I only included those two tests to make clear what is already happening, as this case has probably not been considered before, and to ensure future changes do not change behaviour undetected.

What I **am** intending with this patch is to fix a bug in this for loop in `StaticVersionComparator`:

    for (; i < parts1.length && i < parts2.length; i++) {
        String part1 = parts1[i];
        String part2 = parts2[i];

        Long numericPart1 = numericParts1[i];
        Long numericPart2 = numericParts2[i];

        boolean is1Number = numericPart1 != null;
        boolean is2Number = numericPart2 != null;

        if (part1.equals(part2)) {
            continue;
        }
        if (is1Number && !is2Number) {
            return 1;
        }
        if (is2Number && !is1Number) {
            return -1;
        }
        if (is1Number && is2Number) {
            return numericPart1.compareTo(numericPart2);
        }
        // both are strings, we compare them taking into account special meaning
        Integer sm1 = SPECIAL_MEANINGS.get(part1.toLowerCase(Locale.US));
        Integer sm2 = SPECIAL_MEANINGS.get(part2.toLowerCase(Locale.US));
        if (sm1 != null) {
            sm2 = sm2 == null ? 0 : sm2;
            return sm1 - sm2;
        }
        if (sm2 != null) {
            return -sm2;
        }
        return part1.compareTo(part2);
    }


Since the strings part1 and part 2 are not the same if one has leading zeros, `if (part1.equals(part2))` will not trigger its `continue;` block.
The numeric parts have been parsed by `Longs.tryParse` in `VersionParser`, which _(quote from docs)_ "Parses the specified string as a signed decimal long value.". Apparently happily gobbling up the leading zeros.
This means that `if (is1Number && is2Number)` is true and we call `return numericPart1.compareTo(numericPart2);` breaking out of the loop as soon as we encounter a version part that differs only in leading zeros. _(e.g. "01" vs "1")_

This means that in the case of `1.01.2.3.4` being compared to `1.1.5.6.7`, we report them as being equal without ever comparing the `2.3.4` parts with `5.6.7`.

This results in the version sorter not being able to determine the latest version in a repo that mixes different numbers of leading zeros.
I would agree that mixing these is **not** a good idea, but I encountered the issue depending on an artifact repository, where somebody by mistake had forgotten a leading zero in some versions.

That repo had versions something like these:

  V01.02.0001
  V1.03.0001
  V1.03.0002
  V1.03.0003
  V01.04.0001
  V01.04.0002
  V01.04.0003

And when my version selector specified `V+` I got `V01.02.0001`.
Selecting for `V01+` didn't help. Neither did `+`.
I had to choose `V01.04+` to get the 'real' latest version.
I spent quite a long time trying to deduce the behaviour of Gradle's version selector, to understand why I got `V01.02.0001` instead of `V01.04.0003` until finally giving up and I downloaded Gradle's source and found the bug.

By introducing the `if (numericPart1 == numericPart2)` check, I accept that `01` equals `1` and let the comparator continue with the remaining parts, if any.

    if (is1Number && is2Number) {
        if (numericPart1 == numericPart2) {
            continue;
        }
        return numericPart1.compareTo(numericPart2);
    }


One can argue, and I would whole-heartedly agree, that mixing `V01` and `V1` is a bug in itself, and should be rectified if at all possible.
But I also think this particular bug-fix should be included in Gradle. So Gradle's behaviour is well determined.

__If__ you would prefer `01` **not** to equal `1`, as it does today, `continue;` could easily be replaced with a comparison of the strings again. If so, a decision would have to be made whether more zeros means greater value (`001 > 01 > 1`) or lexicographical sorting is preferred (`1 > 01 > 001`).
Thinking about it while writing this, I've come to feel that `001 == 01 == 1` is the simplest to understand, and it keeps unintentional mixing like the one above from breaking. IMHO, it also makes it easier to grok what would happen if two digits were to not be enough, when going from 01, 02, ... 99, 100, 101, ....

melix commented:
Got it @brouer , I read too fast, and I didn't even realize Gradle considered `1.01` and `1.10` equal. This is also why I dislike when libraries go from `1.9` to `1.10` because the semantics become blurry. I can very easily imagine a producer starting at `1.01`, then issuing `1.02` up to `1.10`... Anyway I agree this is unrelated to this PR.

ljacomet commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/strategy/StaticVersionComparator.java:
```diff
@@ -66,6 +66,9 @@ public int compare(Version version1, Version version2) {
                 return -1;
             }
             if (is1Number && is2Number) {
+                if (numericPart1 == numericPart2) {
```
This may not be true even if the numerical values are the same since it tests for instance equality. I would recommend doing the `compareTo` and a `continue` if `0`, return the result otherwise.

brouer replied:
Right. My Java is a bit rusty.

Fixed now.
